Mathlib/Order/Cover.lean

```diff
@@ -567,6 +567,137 @@ theorem covBy_iff : x ⋖ y ↔ x.1 ⋖ y.1 ∧ x.2 = y.2 ∨ x.2 ⋖ y.2 ∧ x.
+-/
+lemma wcovBy_iff' [Nonempty ι] :
+    a ⩿ b ↔ ∃ i, a i ⩿ b i ∧ ∀ j ≠ i, AntisymmRel (· ≤ ·) (a j) (b j) := by
+  constructor
+  · intro h
+    obtain ⟨i, hi⟩ := exists_forall_antisymmRel_of_wcovBy h
+    exact ⟨i, h.eval _, hi⟩
+  rintro ⟨i, hab, h⟩
+  refine ⟨fun j ↦ (eq_or_ne j i).elim (· ▸ hab.1) (h j · |>.1), fun c hac hcb ↦ ?_⟩
+  have haci : a i < c i := by
+    obtain ⟨hac, j, hj⟩ := Pi.lt_def.1 hac
+    exact (eq_or_ne j i).elim (· ▸ hj) fun hj' ↦
+      ((lt_of_antisymmRel_of_lt (h j hj').symm hj).not_le (hcb.le j)).elim
+  have hcbi : c i < b i := by
+    obtain ⟨hcb, j, hj⟩ := Pi.lt_def.1 hcb
+    exact (eq_or_ne j i).elim (· ▸ hj) fun hj' ↦
+      ((lt_of_lt_of_antisymmRel hj (h j hj').symm).not_le (hac.le j)).elim
+  exact hab.2 haci hcbi
+
+/--
+A characterisation of the `CovBy` relation in products of preorders. See `Pi.covBy_iff` for the
+(more common) version in products of partial orders.
+-/
+lemma covBy_iff' : a ⋖ b ↔ ∃ i, a i ⋖ b i ∧ ∀ j ≠ i, AntisymmRel (· ≤ ·) (a j) (b j) := by
+  constructor
+  · intro h
+    obtain ⟨j, hj⟩ := (Pi.lt_def.1 h.1).2
+    have : Nonempty ι := ⟨j⟩
+    obtain ⟨i, hi⟩ := exists_forall_antisymmRel_of_wcovBy h.wcovBy
+    obtain rfl : i = j := by_contra fun this ↦ (hi j (Ne.symm this)).2.not_lt hj
+    exact ⟨i, covBy_iff_wcovBy_and_lt.2 ⟨h.wcovBy.eval i, hj⟩, hi⟩
+  rintro ⟨i, hi, h⟩
+  have : Nonempty ι := ⟨i⟩
+  refine covBy_iff_wcovBy_and_lt.2 ⟨wcovBy_iff'.2 ⟨i, hi.wcovBy, h⟩, ?_⟩
+  exact Pi.lt_def.2 ⟨fun j ↦ (eq_or_ne j i).elim (· ▸ hi.1.le) (h j · |>.1), i, hi.1⟩
+
+end Preorder
+
+section PartialOrder
+
+variable {ι : Type*} {α : ι → Type*} [∀ i, PartialOrder (α i)] {a b : (i : ι) → α i}
+
+lemma exists_forall_eq_of_covBy (h : a ⋖ b) : ∃ i, ∀ j ≠ i, a j = b j := by
+  obtain ⟨i, hi⟩ := exists_forall_antisymmRel_of_covBy h
+  exact ⟨i, fun j hj ↦ AntisymmRel.eq (hi _ hj)⟩
+
+lemma exists_forall_eq_of_wcovBy [Nonempty ι] (h : a ⩿ b) : ∃ i, ∀ j ≠ i, a j = b j := by
+  obtain ⟨i, hi⟩ := exists_forall_antisymmRel_of_wcovBy h
+  exact ⟨i, fun j hj ↦ AntisymmRel.eq (hi _ hj)⟩
+
+lemma wcovBy_iff [Nonempty ι] : a ⩿ b ↔ ∃ i, a i ⩿ b i ∧ ∀ j ≠ i, a j = b j := by
+  simp [wcovBy_iff']
+
+lemma covBy_iff : a ⋖ b ↔ ∃ i, a i ⋖ b i ∧ ∀ j ≠ i, a j = b j := by
+  simp [covBy_iff']
+
+lemma wcovBy_iff_exists_right_eq [Nonempty ι] [DecidableEq ι] :
+    a ⩿ b ↔ ∃ i x, a i ⩿ x ∧ b = Function.update a i x := by
+  rw [wcovBy_iff]
+  constructor
+  · rintro ⟨i, hi, h⟩
+    exact ⟨i, b i, hi, by simpa [Function.eq_update_iff, eq_comm] using h⟩
+  · rintro ⟨i, x, h, rfl⟩
+    exact ⟨i, by simpa +contextual⟩
+
+lemma covBy_iff_exists_right_eq [DecidableEq ι] :
+    a ⋖ b ↔ ∃ i x, a i ⋖ x ∧ b = Function.update a i x := by
+  rw [covBy_iff]
+  constructor
+  · rintro ⟨i, hi, h⟩
+    exact ⟨i, b i, hi, by simpa [Function.eq_update_iff, eq_comm] using h⟩
+  · rintro ⟨i, x, h, rfl⟩
+    exact ⟨i, by simpa +contextual⟩
+
+lemma wcovBy_iff_exists_left_eq [Nonempty ι] [DecidableEq ι] :
```

Comment:
Can you get `left` versions from `OrderDual`? Or do you consider it to be defeq abuse?

Reply:
I think it is defeq abuse, but more importantly it seems quite a bit more painful to do: there's juggling between the order dual on each of the component types and on the pi type, and commuting the toDual/ofDual with Function.update; overall I don't think it's worth it.